crates/cargo-bsp/src/cargo_communication/cargo_types/cargo_command.rs

```diff
@@ -52,20 +71,160 @@ impl CreateCommand for TestParams {
+                    enabled_features: Default::default(),
+                }),
+                "test_uri2" => Some(TargetDetails {
+                    name: "test_lib1".to_string(),
+                    kind: CargoTargetKind::Lib,
+                    package_name: "test_package2".to_string(),
+                    default_features_disabled: true,
+                    enabled_features: mock_features,
+                }),
+                _ => None,
+            }
+        }
+    }
+
+    fn default_targets() -> Vec<BuildTargetIdentifier> {
+        vec![
+            BuildTargetIdentifier {
+                uri: "test_uri1".to_string(),
+            },
+            BuildTargetIdentifier {
+                uri: "test_uri2".to_string(),
+            },
+        ]
+    }
+
+    #[test]
+    fn test_compile_params_create_command() {
+        let mock_workspace = Arc::new(MockProjectWorkspace);
+        let compile_params = CompileParams {
+            origin_id: None,
+            targets: default_targets(),
+            arguments: vec!["--arg1".to_string(), "--arg2".to_string()],
+        };
+
+        let cmd = compile_params.create_command(mock_workspace);
+        let args: Vec<&OsStr> = cmd.get_args().collect();
+
+        assert_eq!(
+            args,
+            &[
+                "build",
+                "--quiet",
+                "--message-format=json",
+                "--package test_package1 --bin test_bin1  --package test_package2 --lib test_lib1 --feature test_feature1",
+                "--",
+                "--arg1",
+                "--arg2"
+            ]
+        );
+    }
+
+    #[test]
+    fn test_run_params_create_command() {
+        let mock_workspace = Arc::new(MockProjectWorkspace);
+        let run_params = RunParams {
+            origin_id: None,
+            target: BuildTargetIdentifier {
+                uri: "test_uri1".to_string(),
+            },
+            arguments: vec!["--arg1".to_string(), "--arg2".to_string()],
+            data_kind: None,
+            data: None,
+        };
+
+        let cmd = run_params.create_command(mock_workspace);
+        let args: Vec<&OsStr> = cmd.get_args().collect();
+
+        assert_eq!(
+            args,
+            &[
+                "run",
+                "--package test_package1 --bin test_bin1 ",
+                "--message-format=json",
+                "--",
+                "--arg1",
+                "--arg2"
+            ]
+        );
+    }
+
+    #[test]
+    fn test_test_params_create_command() {
+        let mock_workspace = Arc::new(MockProjectWorkspace);
+        let test_params = TestParams {
+            origin_id: None,
+            targets: default_targets(),
+            arguments: vec!["--arg1".to_string(), "--arg2".to_string()],
+            data_kind: None,
+            data: None,
+        };
+
+        let cmd = test_params.create_command(mock_workspace);
+        let args: Vec<&OsStr> = cmd.get_args().collect();
```

Comment:
logika tych funkcji znowu się powtarza... może dałoby się ją jakoś wyodrębnić? Używając insta oczywiście bez sensu, ale nie wiem czy uda się go faktycznie wykorzystać i to byłaby jakaś alternatywa - można by też użyć https://docs.rs/test-case/3.1.0/test_case/

Reply:
Uzylem insta